src/rust/grapl-crypto/src/aead/xchacha_blake3.rs

```diff
@@ -0,0 +1,292 @@
+use blake3::Hash;
+use chacha20::{Key, XNonce, XChaCha12};
+use chacha20::cipher::{NewCipher, StreamCipher};
+use chacha20::cipher::errors::LoopError;
+use crate::hasher::Hasher;
+
+// These domains are used to derive two separate keys from the same origin key.
+// The values can be arbitrary, but must be distinct.
+const DOMAIN_ENCRYPT: &'static [u8] = b"Grapl-12211946";
+const DOMAIN_AUTH: &'static [u8] = b"Grapl-02142019";
+
+const KEY_SIZE: usize = 32;
+const ENC_NONCE_SIZE: usize = 24;
+const MIN_KEY_NONCE_SIZE: usize = 24;
+const TOTAL_NONCE_SIZE: usize = ENC_NONCE_SIZE + MIN_KEY_NONCE_SIZE;
+
+// This aead construct is XChaCha12 with BLAKE3
+// * XChaCha12
+// * BLAKE3
+
+// Inspiration:
+// https://mccarty.io/cryptography/2021/11/29/chacha20-blake3.html
+// https://github.com/soatok/experimental-caead
+// https://soatok.blog/2020/09/09/designing-new-cryptography-for-non-standard-threat-models/
+// https://eprint.iacr.org/2019/1492.pdf
```

Comment:
Should these comments be elsewhere? Either attached to a struct / trait / etc., or for the module as a whole?

(Also, using `///` instead of `//` would ensure they show up in Rust docs.)